### PR TITLE
Promote ExchangeStep to a table and internal entity type.

### PR DESCRIPTION
**Pull request by efoxepstein**
Previously, the plan was to just infer ExchangeSteps, but it seems
simpler and more performant to just explicitly represent them.

---
This change is [Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/42)

### CODE REVIEW

File: src/main/kotlin/org/wfanet/measurement/kingdom/deploy/gcloud/spanner/kingdom.sdl [hunk 1]
```diff
@@ -384,15 +385,35 @@ CREATE TABLE Exchanges (
...
+) PRIMARY KEY (RecurringExchangeId, Date, StepIndex),
+  INTERLEAVE IN PARENT Exchanges ON DELETE CASCADE;
+
+CREATE NULL_FILTERED INDEX ExchangeStepsByModelProviderId
+  ON ExchangeSteps(ModelProviderId, State);
+
+CREATE NULL_FILTERED INDEX ExchangeStepsByDataProviderId
+  ON ExchangeSteps(DataProviderId, State);
+
 CREATE TABLE ExchangeStepAttempts (
```
Comment (yunyeng) [inline src/main/kotlin/org/wfanet/measurement/kingdom/deploy/gcloud/spanner/kingdom.sdl:417]: Do we not have to store ExchangeStepAttempt.State in the table?

File: src/main/kotlin/org/wfanet/measurement/kingdom/deploy/gcloud/spanner/kingdom.sdl [hunk 2]
```diff
@@ -384,15 +385,35 @@ CREATE TABLE Exchanges (
...
   RecurringExchangeId  INT64 NOT NULL,
   Date                 DATE NOT NULL,
```
Comment (yunyeng) [inline src/main/kotlin/org/wfanet/measurement/kingdom/deploy/gcloud/spanner/kingdom.sdl:419]: Why Date not Timestamp ? I am assuming there will be multiple attempts in the same date.

**Review by yunyeng — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/42)* status: 0 of 6 files reviewed, all discussions resolved (waiting on @stevenwarejones)

**Review by stevenwarejones — CHANGES_REQUESTED** [review]
> Reviewed 5 of 6 files at r1.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/42)* status: 5 of 6 files reviewed, 2 unresolved discussions (waiting on @efoxepstein)

___
*[src/main/proto/wfa/measurement/internal/kingdom/exchange_step_attempt.proto, line 60 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/42#-MZnn7A9AmoQjOuhpCB5:-MZnn7A9AmoQjOuhpCB6:b-5j98oz) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/28d290e715c3f0e8f9ac723c187882457d863078/src/main/proto/wfa/measurement/internal/kingdom/exchange_step_attempt.proto#L60)):*
> ```protobuf
>     TEMPORARILY_FAILED = 3;
> 
>     // The attempt has failed permanently and the step should not be retried.
> ```

what do you think instead to call it `UP_FOR_RETRY` .... `TEMPORARILY FAILED` does seem to actively communicate what I think you want it to communicate.
___
*[src/test/kotlin/org/wfanet/measurement/kingdom/service/api/v2alpha/ExchangeStepAttemptsServiceTest.kt, line 233 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/42#-MZnnSh0Amf7J8b3CVXN:-MZnnSh0Amf7J8b3CVXO:bz5dy87) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/28d290e715c3f0e8f9ac723c187882457d863078/src/test/kotlin/org/wfanet/measurement/kingdom/service/api/v2alpha/ExchangeStepAttemptsServiceTest.kt#L233)):*
> ```Kotlin
>             stepIndex = STEP_INDEX
>             attemptNumber = ATTEMPT_NUMBER
>             state = InternalExchangeStepAttempt.State.TEMPORARILY_FAILED
> ```

see other comment about `TEMPORARILY FAILED`

**Review by efoxepstein — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/42)* status: 5 of 6 files reviewed, 2 unresolved discussions (waiting on @stevenwarejones)

___
*[src/main/proto/wfa/measurement/internal/kingdom/exchange_step_attempt.proto, line 60 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/42#-MZnn7A9AmoQjOuhpCB5:-MZnsuCE3A6uuyLBewA-:b6h4rxl) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/28d290e715c3f0e8f9ac723c187882457d863078/src/main/proto/wfa/measurement/internal/kingdom/exchange_step_attempt.proto#L60)):*
<details><summary><i>Previously, stevenwarejones (Steven Ware Jones) wrote…</i></summary><blockquote>

what do you think instead to call it `UP_FOR_RETRY` .... `TEMPORARILY FAILED` does seem to actively communicate what I think you want it to communicate.
</blockquote></details>

To be consistent with github.com/world-federation-of-advertisers/cross-media-measurement-api/pull/21, I'll change this to FAILED and FAILED_STEP. Sound good?
___
*[src/test/kotlin/org/wfanet/measurement/kingdom/service/api/v2alpha/ExchangeStepAttemptsServiceTest.kt, line 233 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/42#-MZnnSh0Amf7J8b3CVXN:-MZntMME44EYcGDbunOI:bsjte78) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/28d290e715c3f0e8f9ac723c187882457d863078/src/test/kotlin/org/wfanet/measurement/kingdom/service/api/v2alpha/ExchangeStepAttemptsServiceTest.kt#L233)):*
<details><summary><i>Previously, stevenwarejones (Steven Ware Jones) wrote…</i></summary><blockquote>

see other comment about `TEMPORARILY FAILED`
</blockquote></details>

See above comment.

**Review by stevenwarejones — APPROVED** [review]
> Reviewed 1 of 6 files at r1.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/42)* status: 3 of 6 files reviewed, 1 unresolved discussion (waiting on @efoxepstein and @stevenwarejones)

___
*[src/main/kotlin/org/wfanet/measurement/kingdom/service/api/v2alpha/ExchangeStepAttemptsService.kt, line 163 at r2](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/42#-MZo9ITr9heeInHCQZO0:-MZo9ITr9heeInHCQZO1:b-o4vjjv) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/dca3445b228e5fafe46c30eca18ebe485e8718c3/src/main/kotlin/org/wfanet/measurement/kingdom/service/api/v2alpha/ExchangeStepAttemptsService.kt#L163)):*
> ```Kotlin
>     // TODO: update public API version to support new [ExchangeStepAttempt] states.
>     InternalExchangeStepAttempt.State.FAILED -> ExchangeStepAttempt.State.FAILED
>     InternalExchangeStepAttempt.State.FAILED_STEP -> ExchangeStepAttempt.State.FAILED
> ```

i like this better

**Pull request merged**